modyn/evaluator/internal/grpc/evaluator_grpc_servicer.py

```diff
@@ -0,0 +1,223 @@
+        self._metric_manager = MetricManager()
+
+        self._model_storage_stub = EvaluatorGRPCServicer.connect_to_model_storage(self._model_storage_address)
+
+    @staticmethod
+    def connect_to_model_storage(model_storage_address: str) -> ModelStorageStub:
+        model_storage_channel = grpc.insecure_channel(model_storage_address)
+        assert model_storage_channel is not None
+        if not grpc_connection_established(model_storage_channel):
+            raise ConnectionError(
+                f"Could not establish gRPC connection to model storage at address {model_storage_address}."
+            )
+        return ModelStorageStub(model_storage_channel)
+
+    def evaluate_model(self, request: EvaluateModelRequest, context: grpc.ServicerContext) -> EvaluateModelResponse:
+        logger.info("Received evaluate model request.")
+
+        if not hasattr(dynamic_module_import("modyn.models"), request.model_id):
+            logger.error(f"Model {request.model_id} not available!")
+            return EvaluateModelResponse(evaluation_started=False)
+
+        fetch_request = FetchModelRequest(model_id=request.trained_model_id)
+        fetch_resp: FetchModelResponse = self._model_storage_stub.FetchModel(fetch_request)
+
+        if not fetch_resp.success:
+            logger.error(
+                f"Trained model {request.trained_model_id} cannot be fetched from model storage. "
+                f"Evaluation cannot be started."
+            )
+            return EvaluateModelResponse(evaluation_started=False)
+
+        with self._lock:
+            evaluation_id = self._next_evaluation_id
+            self._next_evaluation_id += 1
+
+        local_model_path = self._base_dir / f"trained_model_{evaluation_id}.modyn"
+
+        download_file(
+            hostname=self._config["model_storage"]["hostname"],
+            port=int(self._config["model_storage"]["ftp_port"]),
+            user="modyn",
+            password="modyn",
+            remote_file_path=pathlib.Path(fetch_resp.model_path),
+            local_file_path=local_model_path,
+            callback=get_pretrained_model_callback(logger),
+        )
+
+        logger.info(f"Successfully downloaded trained model {request.trained_model_id} to {local_model_path}.")
+
+        metrics = self._setup_metrics(request.metrics)
+        evaluation_info = EvaluationInfo(request, evaluation_id, self._storage_address, metrics, local_model_path)
+        self._evaluation_dict[evaluation_id] = evaluation_info
+
+        exception_queue: mp.Queue[str] = mp.Queue()  # pylint: disable=unsubscriptable-object
+        status_query_queue: mp.Queue[str] = mp.Queue()  # pylint: disable=unsubscriptable-object
+        status_response_queue: mp.Queue[dict[str, Any]] = mp.Queue()  # pylint: disable=unsubscriptable-object
+
+        process = mp.Process(
+            target=evaluate,
+            args=(
+                self._evaluation_dict[evaluation_id],
```

Comment:
I am not sure this approach works. it might work on linux devices with fork but e.g. on mac where we use spawn this is a problem. this dict is not shared between the processes. So i think we need to use a queue into which the evaluator pushes the results upon termination, and then pop from that queue when obtaining the final evaluation results

Reply:
You are totally correct! As discussed, I changed this in the next commit to a queue. However, I do not know if other solutions might be more suitable (like shared memory?).